draft-bar-cfrg-spake2plus.md

```diff
@@ -193,60 +198,72 @@ attackers.
 
 When executing SPAKE2+, A selects x uniformly at random from the
 numbers in the range [0, p), and lets X=x\*P+w0\*M, then transmits pA=X to
-B. Upon receipt of X, A computes h\*X and aborts if the result is equal
+B. Upon receipt of X, B computes h\*X and aborts if the result is equal
 to I. B then selects y uniformly at random from the numbers in [0, p),
-then computes Y=y\*P+w0\*N, and transmits pB=Y to A.
+then computes Y=y\*P+w0\*N, and transmits pB=Y to A. Upon receipt of Y,
+A computes h\*Y and aborts if the result is equal to I.
 
-A computes Z as h\*x\*(Y-w0\*N), and V as h\*w1\*(Y-w0\*N). B computes Z as h\*y\*(X-
-w0\*M) and V as h\*y\*L. Both share Z and V as common values. It is essential
-that both Z and V be used in combination with the transcript to
+A computes Z as h\*x\*(Y-w0\*N), and V as h\*w1\*(Y-w0\*N). B computes Z as
+h\*y\*(X-w0\*M) and V as h\*y\*L. Both share Z and V as common values.
+It is essential that both Z and V be used in combination with the transcript to
 derive the keying material. The protocol transcript encoding is shown below.
 
 ~~~
-TT = len(A) || A || len(B) || B || len(X) || X
-  || len(Y) || Y || len(Z) || Z || len(V) || V
+TT = len(Context) || Context ||
+  || len(A) || A || len(B) || B
+  || len(M) || M || len(N) || N
+  || len(X) || X || len(Y) || Y
+  || len(Z) || Z || len(V) || V
   || len(w0) || w0
 ~~~
 
+Context is an application-specific customization string shared between both
+parties and SHOULD precede the remaining transcript. It might contain the
+name and version number of the higher-level protocol, or simply the name and version
+number of the application. The context MAY include additional data such as the
+chosen ciphersuite and PBKDF parameters like the iteration count or salt.
+The context and its length prefix MAY be omitted.
+
 If an identity is absent, it is omitted from the transcript entirely. For example,
-if both A and B are absent, then TT = len(X) || X || len(Y) || Y || len(Z) || Z || len(w0) || w0.
-Likewise, if only A is absent, TT = len(B) || B || len(X) || X || len(Y) || Y || len(Z) || Z || len(w0) || w0.
+if both A and B are absent, then TT = len(Context) || Context || len(M) || M || len(N) || N || len(X) || X || len(Y) || Y || len(Z) || Z || len(w0) || w0.
+Likewise, if only A is absent, TT = len(Context) || Context || len(B) || B || len(M) || M || len(N) || N || len(X) || X || len(Y) || Y || len(Z) || Z || len(w0) || w0.
```

Comment:
This is worrisome :-( The security proofs assume treat Hash as a RO, which is the case if the input is prefix-free, but how do we guarantee prefix-freedom if some values are optional? As an example, consider two different protocols Foo and Bar. In Foo, A's identity is omitted. In Bar, B's identity is omitted. Moreover, assume that there exists a party acting as B in Foo with identity (name) N and a party acting as A in Bar with identity (name) N'. If N = N', and if the context string is omitted from both protocols, we can have a RO collision. 

Is this a problem? (I think an easy fix is to add a type byte before each length value in the transcript, e.g.:

```
TT = 0x00 || len(Context) || Context ||
  0x01 || || len(A) || A || 0x02 || len(B) || B
  0x03 || || len(M) || M || 0x04 || len(N) || N
  0x05 || len(X) || X || 0x06 || len(Y) || Y
  0x07 || len(Z) || Z || 0x08 || len(V) || V
  0x09 || len(w0) || w0
```

Cumbersome, and probably a hard change to pitch, but I think this is probably needed.

Reply:
(Alternatively, we can remove the "optional" part and just always encode the lengths. That'd probably be simpler.)

Reply:
That's a great point, thanks for bringing this up! I've removed the "optional" part and identities are now mandatory parts of the transcript, even if their lengths are zero and the identities themselves are empty octet strings. It even simplifies things a bit. 👍